k9mail/src/main/java/com/fsck/k9/helper/AppUpdater.kt

```diff
@@ -50,33 +57,60 @@ class AppUpdater(private val context: Context, private val cacheDir: File) {
     private fun updateAppSettingsAndVersion(oldVersion: Long, newVersion: Long, context: Context) {
         val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
         scope.launch {
-            val editor = Preferences.getPreferences(context).storage.edit()
-            updateAppSettings(oldVersion, editor)
+            val storage = Preferences.getPreferences(context).storage
+            val editor = storage.edit()
+            updateAppSettings(oldVersion, newVersion, storage, editor)
             K9.setAppVersionCode(newVersion) // version code updated every time
             K9.save(editor)
             editor.commit()
         }
     }
 
-    private fun updateAppSettings(oldVersion: Long, editor: StorageEditor) {
-        when (oldVersion) {
-            v317 -> {
-                // cleanup removed settings
-                editor.remove("messageViewArchiveActionVisible")
-                editor.remove("messageViewDeleteActionVisible")
-                editor.remove("messageViewMoveActionVisible")
-                editor.remove("messageViewCopyActionVisible")
-                editor.remove("messageViewSpamActionVisible")
-            }
-            else -> {}
+    private fun updateAppSettings(
+        oldVersion: Long,
+        newVersion: Long,
+        storage: Storage,
+        editor: StorageEditor
+    ) {
+        fun needsUpdateFor(target: Long) = needsUpdate(target, oldVersion, newVersion)
```

Comment:
where is the oldVersion stored? What if I update skipping some versions?

Reply:
I am checking if the target version is in the range oldVersion <= target < newVersion.

Actually thanks for the comment, I was doing oldVersion <= target <= newVersion. I fixed it and added the unit test.

The old version is stored in the storage. This was a great comment as well, as it results that because of doing the updater operations now earlier in the app initialization process, the data we get from the app itself was only default and we need to take it from the storage.

I also noticed the same applies when saving the data. We don't want to save all app data because it is not loaded yet at that early moment so instead we **only save the data we changed to the disk**.

Reply:
perfect, it was good to check it then :)